### PR TITLE
[#2490] Functional Exchange looping on the same function

### LINKED ISSUE #2490: Functional Exchange looping on the same function
- A FE cannot be created between two ports on the same function using the diagram interactor, yet using a drag and drop, or a direct edition using the expert tab of the property box allows it
- When we create this looping FE, it disappears from the diagram
- There isn't a validation to detect this pattern in the model

### PR DESCRIPTION
**Pull request by irina-toma**
Resolves #2490

Change-Id: Idc8d7cfcd3e62c2508b17348642ddf08d6fe3fe4
Signed-off-by: Irina Toma <irina_toma@rocketmail.com>

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml [hunk 1]
```diff
@@ -114,6 +114,25 @@
                                This rule generates a warning if a Function Port and all its connected Functional Exchanges do not have the exact same Exchange Items.
                             </description>
                          </constraint>
+                         <constraint
+                               class="org.polarsys.capella.core.data.fa.validation.functionalExchange.FunctionalExchange_loopsOnSameFunction"
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml:118]: Can you also update the embedded documentation
https://github.com/eclipse/capella/tree/master/doc/plugins/org.polarsys.capella.validation.doc

File: core/plugins/org.polarsys.capella.core.data.fa.validation/src/org/polarsys/capella/core/data/fa/validation/functionalExchange/FunctionalExchange_loopsOnSameFunction.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * This rule ensures that there are no Functional Exchanges defined with input and output port on the same function
+ */
+public class FunctionalExchange_loopsOnSameFunction extends AbstractValidationRule {
+
+  /**
+   * @see org.eclipse.emf.validation.AbstractModelConstraint#validate(org.eclipse.emf.validation.IValidationContext)
+   */
+  @Override
+  public IStatus validate(IValidationContext ctx) {
+    // Raise a warning if one of exchange item convoyed by related functional exchanges are not convoyed by the port
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.validation/src/org/polarsys/capella/core/data/fa/validation/functionalExchange/FunctionalExchange_loopsOnSameFunction.java:35]: comment not uptodate/to remove

File: core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml [hunk 3]
```diff
@@ -114,6 +114,25 @@
...
+                               id="DCOM_23"
+                               isEnabledByDefault="true"
+                               lang="java"
+                               mode="Batch"
+                               name="DCOM_23 - Loops on same function check"
+                               severity="ERROR"
+                               statusCode="1">
+                            <message>
+                               {0} (Functional Exchange) {1} loops on same {2} {3}.
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml:127]: on operational analysis, it's called an Interaction, can you adapt message accordingly

File: core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml [hunk 4]
```diff
@@ -114,6 +114,25 @@
...
+                            </message>
+                            <target
+                                  class="FunctionalExchange">
+                            </target>
+                            <description>
+                               This rule generates an error if the Functional Exchange has the source and target ports defined on the same Function.
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.validation/plugin.xml:133]: can you adapt for operational analysis. there is no ports and function are called operational activity

File: core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/plugin.properties [hunk 5]
```diff
@@ -105,5 +105,14 @@ quickFix.DCOM_20_PortResolver_Target.desc = Show target Function Input Port in P
 quickFix.RemoveUnusedExchangeItems.label = Remove unused Exchange Items
 quickFix.RemoveUnusedExchangeItems.desc = Removes unused Exchange Items from the port
 
+quickFix.DCOM_23_Resolver_Source.desc = Show source Function Output Port in Project Explorer
+quickFix.DCOM_23_Resolver_Source.label = Show source Function Output Port in Project Explorer
+
+quickFix.DCOM_23_Resolver_Target.label = Show target Function Input Port in Project Explorer
+quickFix.DCOM_23_Resolver_Target.desc = Show target Function Input Port in Project Explorer
+
+quickFix.DCOM_23_Resolver_Delete.desc = Delete invalid Functional Exchange
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/plugin.properties:114]: Can you adapt message on Operational analysis level / Interaction

File: core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/src/org/polarsys/capella/core/data/fa/ui/quickfix/resolver/DCOM_23_Resolver_Source.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import org.polarsys.capella.core.data.fa.ui.quicfix.helpers.QuickFixNavigationHelper;
+import org.polarsys.capella.core.validation.ui.ide.quickfix.AbstractCapellaMarkerResolution;
+
+/**
+ * QuickFix allowing to show source Function Input Port in Project Explorer.
+ */
+public class DCOM_23_Resolver_Source extends AbstractCapellaMarkerResolution {
+
+  @Override
+  public void run(IMarker marker) {
```
Comment (pdulth) [inline core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/src/org/polarsys/capella/core/data/fa/ui/quickfix/resolver/DCOM_23_Resolver_Source.java:31]: The two "show" resolver doesn"t apply on Operational analsys layer. There is a AbstractCapellaMarkerResolution.canResolve that might allow to make it not visible on Interactions

**Review by pdulth — CHANGES_REQUESTED** [review]

**Review by pdulth — APPROVED** [review]

**Review by pdulth — CHANGES_REQUESTED** [review]
can you fix failing junit ?
https://ci.eclipse.org/capella/job/capella-product/job/PR-2491/4/testReport/junit/org.polarsys.capella.test.doc.ju/StyleCheckTestCase/StyleCheckTestCase/

To generate the documentation, you need to use the build.xml at the root folder of the plugin o keep the css.

File: core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/plugin.properties [hunk 7]
```diff
@@ -105,5 +105,14 @@ quickFix.DCOM_20_PortResolver_Target.desc = Show target Function Input Port in P
 quickFix.RemoveUnusedExchangeItems.label = Remove unused Exchange Items
 quickFix.RemoveUnusedExchangeItems.desc = Removes unused Exchange Items from the port
 
+quickFix.DCOM_23_Resolver_Source.desc = Show source Function Output Port in Project Explorer
+quickFix.DCOM_23_Resolver_Source.label = Show source Function Output Port in Project Explorer
+
+quickFix.DCOM_23_Resolver_Target.label = Show target Function Input Port in Project Explorer
+quickFix.DCOM_23_Resolver_Target.desc = Show target Function Input Port in Project Explorer
+
+quickFix.DCOM_23_Resolver_Delete.desc = Delete
```
[the remark this answers is not in the document]
Comment (irina-toma) [reply core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/plugin.properties:114]: Michel told me to leave it like that :)

File: core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/src/org/polarsys/capella/core/data/fa/ui/quickfix/resolver/DCOM_23_Resolver_Source.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Override
+  protected boolean quickFixAllSimilarEnabled(Collection<IMarker> markers) {
+    return false;
+  }
+
+
+  @Override
+  protected boolean canResolve(IMarker marker) {
+    final EObject modelElement = getModelElements(marker).get(0);
```
[the remark this answers is not in the document]
Comment (irina-toma) [reply core/plugins/org.polarsys.capella.core.data.fa.ui.quickfix/src/org/polarsys/capella/core/data/fa/ui/quickfix/resolver/DCOM_23_Resolver_Source.java:51]: It doesn't, enabled() removes them from the list

**Review by pdulth — APPROVED** [review]

**Pull request merged**